<div>
<a href="https://www.audiolabs-erlangen.de/fau/professor/mueller"><img src="data_layout/PCP_Teaser.png" width=100% style="float: right;" alt="PCP Teaser"></a>
</div>

# Unit 1: Get Started

<ul>
  <li><a href="#learn" style="color:#E65100;"><strong>Overview and Learning Objectives</strong></a></li>
  <li><a href="#quick_start" style="color:#E65100;"><strong>Quick Start:</strong> Installing and Starting PCP</a></li>
  <li><a href="#obtain_pcp" style="color:black;">Obtaining the PCP Material</a></li>
  <li><a href="#background_git" style="color:#1B5E20;"><strong>Background:</strong> Version Control with Git and GitHub</a></li>
  <li><a href="#conda" style="color:black;">Python, Miniconda, and Conda</a></li>
  <li><a href="#environment_file" style="color:black;">Python Environments and the PCP Environment File</a></li>
  <li><a href="#create_environment" style="color:black;">Creating and Managing the PCP Environment</a></li>
  <li><a href="#background_boxes" style="color:#1B5E20;"><strong>Background:</strong> A Pause and a Change of Perspective</a></li>
  <li><a href="#jupyter" style="color:black;">Using Jupyter Notebooks</a></li>
  <li><a href="#export" style="color:black;">Exporting Notebooks</a></li>
  <li><a href="#exercise_setup" style="color:#006064;"><strong>Exercise 1:</strong> Verifying Your PCP Setup</a></li>
  <li><a href="#exercise_notebook" style="color:#006064;"><strong>Exercise 2:</strong> Your First Notebook Experiment</a></li>
</ul>

<a id="learn"></a>

<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding: 10px 20px;
  margin: 10px 0;
">
<h2 style="color:#E65100; margin-top:0;">
Overview and Learning Objectives
</h2>

<p style="color:#E65100;">
Welcome to the <strong>Preparation Course for Python</strong> (PCP). This first unit helps you obtain the course material, install the required software, and open the PCP notebooks on your computer. Along the way, it introduces the tools involved in this process: <strong>GitHub</strong> stores and publishes the course files, <strong>Miniconda</strong> provides Python and the package manager Conda, a <strong>Python environment</strong> contains the software required for the course, and <strong>Jupyter Notebook</strong> lets you read, edit, and execute notebooks interactively.
</p>

<p style="color:#E65100;">
If you only want to read the material, you can use the <a href="https://www.audiolabs-erlangen.de/PCP">HTML version of PCP</a> directly in a web browser. To execute the code, change parameters, and complete the exercises, you need the executable notebook files and a suitable Python environment. After completing this unit, you should be able to:
</p>

<ul style="color:#E65100;">
  <li>Distinguish between static HTML pages and executable Jupyter notebooks.</li>
  <li>Download the PCP material from GitHub, either as an archive or using Git.</li>
  <li>Explain the roles of Python, Miniconda, Conda, packages, and environments.</li>
  <li>Create, activate, and inspect the PCP environment using <code>environment.yml</code>.</li>
  <li>Start Jupyter Notebook from the PCP directory and open a notebook.</li>
  <li>Work with code and Markdown cells and execute a notebook in a reliable order.</li>
  <li>Export a notebook as a static HTML or PDF document.</li>
</ul>

<p style="color:#E65100;">
The unit contains two exploratory exercises:
</p>

<ul style="color:#E65100;">
  <li><a href="#exercise_setup">Exercise 1</a>: Verify that the PCP environment and its required software are working correctly.</li>
  <li><a href="#exercise_notebook">Exercise 2</a>: Perform a first notebook experiment by executing, modifying, and exporting a simple example.</li>
</ul>

<p style="color:#E65100; margin-bottom:0;">
The main goal is practical: by the end of this unit, the PCP notebooks should be running on your computer. The explanations help you understand each tool well enough to repeat the same workflow for other Python projects.
</p>
</div>

In [1]:
# --- Python standard library ---
import os
import sys

# --- Core scientific stack ---
import matplotlib
import numpy as np
from matplotlib import pyplot as plt

# --- PCP package ---
import libpcp

<a id="quick_start"></a>

<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding: 10px 20px;
  margin: 10px 0;
">
<strong style="color:#E65100; display:block; break-after:avoid; page-break-after:avoid;">Quick Start: Installing and Starting PCP</strong>

<p style="color:#E65100;">
The following steps summarize the complete local setup. If you have worked with Python environments and Jupyter notebooks before, this list may be all you need: the procedure is standard for many scientific Python projects. If terms such as <code>conda</code>, environment, or terminal are new to you, do not worry. We explain the tools and commands one by one in the following sections. By the end of the unit, the instructions below should look considerably less mysterious.
</p>

<p style="color:#E65100;">
The first steps are normally required only once. Activating the environment and starting Jupyter Notebook are repeated whenever you begin a new PCP session. Commands such as <code>conda activate PCP</code> are entered in a <strong>terminal</strong>, a text-based interface for starting programs and managing files. On Windows, use the Anaconda Prompt; on macOS or Linux, use a terminal application. After Jupyter Notebook has started, Python statements are entered into notebook code cells.
</p>

<ol style="color:#E65100; margin-bottom:0;">
  <li>
    Install <a href="https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html"><strong>Miniconda</strong></a>, which provides Python and the package manager Conda.
  </li>
  <li>
    Download the PCP project from <a href="https://github.com/groupmm/PCP"><strong>GitHub</strong></a> and extract the downloaded archive.
  </li>
  <li>
    Open the <strong>Anaconda Prompt</strong> on Windows or a terminal application on macOS or Linux.
  </li>
  <li>
    Change to the directory containing the PCP files:
    <code>cd PCP</code>
  </li>
  <li>
    Create the PCP environment:
    <code>conda env create -f environment.yml</code>
  </li>
  <li>
    Activate the PCP environment:
    <code>conda activate PCP</code>
  </li>
  <li>
    Start Jupyter Notebook:
    <code>jupyter notebook</code>
  </li>
  <li>
    When Jupyter Notebook opens in your browser, select <code>PCP.ipynb</code> to open the course overview.
  </li>
</ol>

</div>

<a id="obtain_pcp"></a>
## Obtaining the PCP Material

PCP is available in two formats. The <a href="https://www.audiolabs-erlangen.de/PCP"><strong>HTML version</strong></a> can be read immediately in a web browser. It contains the explanations, code, figures, and audio examples, but its code cannot be changed or executed. The <a href="https://github.com/groupmm/PCP"><strong>Jupyter notebook version</strong></a> contains executable files with the extension <code>.ipynb</code>. Use this version if you want to run the examples, change parameters, or work on the exercises.

The project is hosted on <a href="https://github.com/groupmm/PCP"><strong>GitHub</strong></a>, an online platform for publishing and developing software projects. A project stored on GitHub is called a <strong>repository</strong>. The PCP repository contains the notebooks together with the required Python modules, images, audio files, and configuration files.

The simplest way to obtain the project is to download an archive:

<ol>
  <li>Open the <a href="https://github.com/groupmm/PCP">PCP repository</a> in a browser.</li>
  <li>Select <strong>Code</strong> and then <strong>Download ZIP</strong>.</li>
  <li>Extract the downloaded archive to a directory on your computer.</li>
  <li>If desired, rename the extracted directory to <code>PCP</code>.</li>
</ol>

GitHub also works with <a href="https://git-scm.com/"><strong>Git</strong></a>, a program that records changes to files and supports collaboration on software projects. If Git is installed, the repository can be copied, or <strong>cloned</strong>, using the following terminal command: <code>git clone https://github.com/groupmm/PCP.git</code>

Cloning creates a directory named <code>PCP</code>. Unlike a downloaded archive, a Git repository retains its version history and can later be updated using Git. No knowledge of Git is required for the remainder of this course; downloading the ZIP archive is sufficient. The following background box provides a broader introduction to Git and GitHub.


<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding: 10px 20px;
  margin: 10px 0;
">
<p style="color:#E65100; margin:0;">
<strong>Note: Running PCP Online.</strong>
PCP can also be run through browser-based services such as <a href="https://colab.research.google.com/">Google Colab</a> and <a href="https://mybinder.org/v2/gh/groupmm/PCP/master">Binder</a>, without installing Python locally. The local setup remains the recommended and tested approach because it provides direct access to all notebooks, modules, and data. Online services may require additional preparation, and their startup time and file handling may differ. Current instructions are provided in the <a href="https://github.com/groupmm/PCP#readme">PCP repository documentation</a>. The remainder of this unit describes the local setup.
</p>
</div>

<a id="conda"></a>

<a id="background_git"></a>

<div style="
  background-color: #C8E6C9A0;
  border: 1px solid #81C784A0;
  padding: 10px 20px;
  margin: 10px 0;
">

<h2 style="color:#1B5E20; margin-top:0;">
Background: Version Control with Git and GitHub
</h2>

<p style="color:#1B5E20;">
Software projects change over time: errors are corrected, explanations are improved, and new functionality is added. Keeping many copies with names such as <code>project_01</code>, <code>project_02</code>, and <code>project_final</code> quickly becomes confusing. A <strong>version-control system</strong> records these changes in an organized way. It shows what has changed, preserves earlier versions, and helps several people work on the same project. <a href="https://en.wikipedia.org/wiki/Git"><strong>Git</strong></a> is one of the most widely used version-control systems in software development and scientific programming. A project managed by Git is called a <strong>repository</strong>. It contains the project files together with a history of the changes recorded by its developers. Git runs locally on your computer and can be used without an internet connection. <a href="https://github.com/"><strong>GitHub</strong></a> is a separate online service that hosts Git repositories and provides additional tools for sharing projects, reporting problems, reviewing changes, and collaborating with others. In short, Git manages the history of a project, while GitHub provides an online place for storing and sharing it. The following table summarizes several important Git commands:
</p>

<table style="
  width:85%;
  border-collapse:collapse;
  table-layout:fixed;
  margin:10px auto;
  color:#1B5E20;
  text-align:left;
  font-size:95%;
">
  <colgroup>
    <col style="width:35%;">
    <col style="width:65%;">
  </colgroup>
  <thead>
    <tr style="background-color:#E8F5E9;">
      <th style="border:1px solid #81C784; padding:4px 6px; text-align:left;">Command</th>
      <th style="border:1px solid #81C784; padding:4px 6px; text-align:left;">Purpose</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git clone URL</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Create a local copy of an online repository.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git status</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Show which files have changed and which changes are selected for the next commit.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git pull</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Download and integrate changes from an online repository.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git add filename</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Select a changed file for the next commit.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git commit -m "Message"</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Record the selected changes locally with a short description.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #81C784; padding:4px 6px;"><code>git push</code></td>
      <td style="border:1px solid #81C784; padding:4px 6px;">Upload local commits to an online repository.</td>
    </tr>
  </tbody>
</table>

<p style="color:#1B5E20;">
Several files can be selected together. For example, <code>git add *.ipynb</code> selects all notebook files in the current directory, while <code>git add --all</code> selects all changes in the repository. Before committing, use <code>git status</code> to check that you selected the intended files.
</p>

<p style="color:#1B5E20; margin-bottom:0;">
Not every user can upload changes directly to every repository. Contributions to a public project are commonly prepared in a personal copy called a <strong>fork</strong> and proposed to the original project through a <strong>pull request</strong>. Each project may define its own contribution guidelines, which should be read before submitting changes. Git and GitHub are not required for completing PCP, but they become valuable when you want to update a project, keep track of your own work, or collaborate with others. To continue learning, the official <a href="https://git-scm.com/docs/gittutorial">Git tutorial</a> introduces the basic Git workflow, while the beginner-friendly <a href="https://docs.github.com/en/get-started/using-github/hello-world">GitHub Hello World guide</a> explains repositories, branches, commits, and pull requests through a small practical example.
</p>

</div>

<a id="conda"></a>
## Python, Miniconda, and Conda

The <strong>Preparation Course for Python</strong> introduces Python through examples from scientific computing, mathematics, and signal processing. Before we begin learning the language in <a href="PCP_02_python.html">Unit 2</a>, we first prepare the software needed to run the examples and complete the exercises. This includes Python itself and several additional packages used throughout the course. The basic Python language does not provide every tool needed for scientific computing. Additional collections of code, called <strong>packages</strong>, extend its functionality. PCP uses packages such as **NumPy** for numerical calculations, **Matplotlib** for visualization, and **Jupyter Notebook** for interactive computing. Some packages require other packages in order to work. These requirements are called <strong>dependencies</strong>.

Installing Python and managing all required packages individually would be inconvenient. We therefore use <a href="https://docs.conda.io/"><strong>Conda</strong></a>, a tool for managing packages and Python environments. Conda determines which dependencies are required, installs suitable versions, and keeps the software used by different projects separate. <a href="https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html"><strong>Miniconda</strong></a> is a small distribution that provides Python, Conda, and a few essential components. It gives us everything needed to create the PCP environment without installing a large collection of unrelated packages. To install Miniconda:

<ol>
  <li>Open the official <a href="https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html">Conda installation page</a>.</li>
  <li>Select the Miniconda installer for your operating system.</li>
  <li>Run the installer and follow its instructions.</li>
  <li>On Windows, open the <strong>Anaconda Prompt</strong>. On macOS or Linux, open a terminal application.</li>
  <li>Verify the installation by entering: <code>conda --version</code></li>
</ol>

If Conda is installed correctly, the final command displays its version number. The exact number may differ between computers. We can now use Conda to create a separate software environment for PCP.

<a id="environment_file"></a>
## Python Environments and the PCP Environment File

A Python project rarely uses Python alone. It may depend on many packages, and these packages continue to change as developers correct errors and add new features. Different projects may therefore require different versions of the same package. If all packages are installed together, updating the software for one project may cause another project to stop working. A <strong>Python environment</strong> keeps the software for one project separate from the software used by other projects. You can think of it as a dedicated toolbox containing a particular Python version and a particular collection of packages. Activating an environment tells the computer which toolbox to use. PCP therefore has its own environment containing the software versions intended for this course.

It is important to distinguish the <strong>project directory</strong> from the <strong>Python environment</strong>. The project directory contains the files you work with, including notebooks, Python modules, images, audio, and data. The environment contains the software used to open and execute those files. Deleting an environment does not delete the project files, and copying the project directory does not automatically install the required software.

The PCP project includes a file named <code>environment.yml</code>. The extension <code>.yml</code> indicates that the file uses <a href="https://yaml.org/"><strong>YAML</strong></a>, a text format for storing structured information in a form that is easy for both people and programs to read. YAML files are often used to describe the settings and software requirements of a project. In this case, <code>environment.yml</code> specifies the name of the PCP environment, the locations from which packages are obtained, and the software required by the course. The following code cell reads the file and displays its current contents. You do not need to understand the Python code yet; its purpose here is simply to show the file.

In [2]:
fn = os.path.join('environment.yml')

with open(fn, 'r', encoding='utf-8') as stream:
    env = stream.read()

print(env)

name: PCP  # the name of the environment

channels:  # conda channels from which packages are downloaded
  - conda-forge
  - nodefaults 

dependencies:
  - python=3.14.*                 # Plain Python
  - pip=26.2.*                    # Package installer for Python
  - numpy=2.5.*                   # NumPy
  - matplotlib=3.11.*             # Matplotlib
  - jupyter=1.1.*                 # Jupyter metapackage (notebook, lab, ipykernel)
  - pip:                          # Packages that are installed via pip
    - nbconvert[webpdf]==7.17.*   # HTML/PDF export
    - nbstripout==0.9.*           # strip notebook output



The first entry gives the environment the name <code>PCP</code>. The <code>channels</code> section tells Conda where packages should be downloaded. In this case, the packages come from <code>conda-forge</code>, a widely used community collection of software packages. The entry <code>nodefaults</code> tells Conda not to search its standard package channels in addition to the listed channel. The <code>dependencies</code> section lists the software to be installed in the environment. An entry such as <code>python=3.14.*</code> specifies both a package and an allowed version range. The asterisk permits any available update within version <code>3.14</code>. The final subsection lists packages installed using <strong>pip</strong>, Python's own package installer. Indentation is part of the structure of a YAML file, so indented entries belong to the section above them.

You do not need to install these packages individually or decide which versions work together. This is exactly what Conda is designed to manage. It reads <code>environment.yml</code>, identifies the requested packages and their dependencies, and searches for a collection of versions that are compatible with one another. This process is called <strong>dependency resolution</strong>. Conda then downloads the selected packages and installs them together in the new environment. A <strong>conflict</strong> occurs when no collection of package versions can satisfy all requirements. For example, one package may require an older version of another package, while a second package requires a newer and incompatible version. In this situation, Conda reports the problem instead of silently creating an environment whose packages cannot work together. Specifying the package versions also supports <strong>reproducibility</strong>: students working on different computers can create closely matching software environments and are therefore more likely to obtain the same results.

<a id="create_environment"></a>
## Creating and Managing the PCP Environment

With Conda installed and <code>environment.yml</code> available, we can now create the PCP environment. Complete the following steps in order:

<ol>
  <li>
    Open the <strong>Anaconda Prompt</strong> on Windows or a terminal application on macOS or Linux. Use <code>cd PCP</code> to enter the directory containing the PCP files. Here, <code>cd</code> means <strong>change directory</strong>. The exact command depends on where you stored the project. You are in the correct directory when it contains both <code>environment.yml</code> and <code>PCP.ipynb</code>.
  </li>
  <li>
    Create the environment using <code>conda env create -f environment.yml</code>. The option <code>-f</code> tells Conda to read the environment settings from the specified file. Conda then installs Python together with the required packages and dependencies. This may take several minutes, but it normally needs to be done <strong>only once</strong>.
  </li>
  <li>
    Activate the environment using <code>conda activate PCP</code>. Activating an environment tells the terminal to use the Python installation and packages prepared for PCP. The terminal prompt will often begin with <code>(PCP)</code>. Repeat this step <strong>whenever you begin a new PCP session</strong>. When you finish working, use <code>conda deactivate</code> to leave the environment.
  </li>
</ol>

The following table summarizes the most useful Conda commands. Some are needed for the regular PCP workflow, while others help you inspect, update, or recreate the environment.

<table style="
  width:85%;
  border-collapse:collapse;
  table-layout:fixed;
  margin:10px auto;
  color:black;
  text-align:left;
  font-size:95%;
">
  <colgroup>
    <col style="width:44%;">
    <col style="width:56%;">
  </colgroup>
  <thead>
    <tr style="background-color:#E3E3E3;">
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Command</th>
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Purpose</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda --version</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Check that Conda is installed and display its version.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda update -n base conda</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Update the Conda program in its main environment, named <code>base</code>.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda env create -f environment.yml</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Create an environment using the settings in <code>environment.yml</code>.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda activate PCP</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Activate the PCP environment for the current terminal session.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda deactivate</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Leave the active environment without deleting it.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda info --envs</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">List the environments available on the computer and indicate which one is active.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda list</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">List the packages installed in the active environment.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda env update -f environment.yml</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Update the environment using the current environment file.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>conda env remove -n PCP</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Remove the environment named <code>PCP</code> and its installed software.</td>
    </tr>
  </tbody>
</table>

<p>
Deactivating and removing an environment are different operations. <code>conda deactivate</code> simply stops using the environment in the current terminal. <code>conda env remove -n PCP</code> deletes the environment and its installed software, but it does not delete the PCP project directory, notebooks, or data. If the environment becomes damaged or difficult to update, you can remove it and create a fresh version from <code>environment.yml</code>.
</p>

<a id="background_boxes"></a>

<div style="
  background-color: #C8E6C9A0;
  border: 1px solid #81C784A0;
  padding: 10px 20px;
  margin: 10px 0;
">
    
<h2 style="color:#1B5E20; margin-top: Pragmatic;">
Background: A Pause and a Change of Perspective
</h2>

<p style="color:#1B5E20;">
Throughout PCP, green <strong>background boxes</strong> offer short excursions from the main course material. Some introduce a larger mathematical or scientific idea. Others connect the current topic with applications, history, or related fields. These <strong>connections</strong> help us see why an idea matters and how it fits into a broader picture. The boxes also offer different levels of depth. Some may require a little more attention, while others share a lighter topic or an interesting fact. Each provides a <strong>change of perspective</strong> from the regular sequence of explanations, formulas, and code. Sometimes a short detour makes the main path easier to understand and remember.
</p>

<p style="color:#1B5E20; margin-bottom:0;">
You may also treat a background box as an invitation to <strong>slow down</strong>. If your attention begins to fade, <strong>take a real break</strong>: stand up, stretch, open a window, take a few steps, go outside for some fresh air, or get a cup of tea or coffee. You do not need to understand everything at once or in the same depth. Begin with the general picture and return later to explore the details. Difficult ideas often become clearer after several encounters. Reading can introduce an idea, but deeper understanding grows when you work with it yourself. Be brave enough to <strong>experiment</strong>: run the code, change parameters, make a prediction, and see what happens. Try ideas that may not work, and allow yourself to <strong>make mistakes</strong>. A surprising result or an error message is often the beginning of a better understanding. We hope these boxes give you useful context, fresh perspectives, and some <strong>breathing room</strong> as you find your own way through PCP.
</p>

</div>

<a id="jupyter"></a>
## Using Jupyter Notebooks

The PCP course is presented through interactive documents called **Jupyter notebooks**. These notebooks are opened and edited using [**Jupyter Notebook**](https://jupyter-notebook.readthedocs.io/en/latest/), an application designed for interactive computing. A notebook is divided into sections called **cells**. A **Markdown cell** contains explanatory text, headings, lists, links, images, and mathematical formulas, while a **code cell** contains executable Python instructions and displays their results directly below the cell. Together, these cells allow explanations, calculations, figures, and audio to appear in a single document. PCP uses this format to place explanations and experiments next to each other. You can read a mathematical idea, execute the corresponding code, change a parameter, and immediately see or hear the result. The examples are therefore not fixed demonstrations. They are starting points that you can explore, modify, and extend with your own ideas.

The PCP project files are now on your computer, and the PCP environment contains the software needed to use them. Open the Anaconda Prompt on Windows or a terminal application on macOS or Linux. Then complete the following steps:

<ol>
  <li>Use <code>cd PCP</code> to change to the PCP project directory.</li>
  <li>Use <code>conda activate PCP</code> to activate the PCP environment.</li>
  <li>Use <code>jupyter notebook</code> to start Jupyter Notebook.</li>
</ol>

The final command normally opens the Jupyter file browser in your default web browser. This page shows the files and directories contained in the PCP project. Select <code>PCP.ipynb</code> to open the course overview, or select one of the unit notebooks directly.

<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding: 10px 20px;
  margin: 10px 0;
">
<p style="color:#E65100; margin:0;">
<strong>Note: Opening PCP in Jupyter.</strong>
Start Jupyter Notebook from the PCP directory. This directory then becomes the starting point of the Jupyter file browser, and the notebooks can find the accompanying Python modules, images, audio files, and data using the expected paths. When navigating between PCP notebooks, follow links labeled <strong>IPYNB</strong>. The filename extension <code>.ipynb</code> comes from <strong>IPython Notebook</strong>, the predecessor of Jupyter Notebook. Jupyter retained this extension when the notebook system was extended beyond Python to other programming languages. Files ending in <code>.ipynb</code> are interactive notebooks whose code cells can be executed, while files ending in <code>.html</code> are static exports that can only be read.
</p>
</div>

A notebook is usually read and executed from top to bottom. Select a cell and press <code>Shift + Enter</code> to process it and continue to the next cell. For a **Markdown cell**, this turns its source text into formatted content. For a **code cell**, it sends the Python instructions to the <strong>kernel</strong>, the running Python process that performs the computations. A number such as <code>[3]</code> beside a code cell indicates when it was executed during the current session. Empty brackets <code>[ ]</code> indicate that the cell has not yet been executed, while <code>[*]</code> indicates that it is still running. The kernel remembers imported packages, variables, functions, and previous results until it is restarted or stopped. This remembered information is called the notebook's **state**. A later cell can use a variable created in an earlier cell, but only if the earlier cell has already been executed. Since cells can be run in any order, the visible order of a notebook does not always match its actual execution order.

<div style="
  background-color: #FFE0B2A0;
  border: 1px solid #FFC671A0;
  padding: 10px 20px;
  margin: 10px 0;
">
<p style="color:#E65100; margin:0;">
<strong>Note: Execution Order Matters.</strong>
A notebook may appear to work because the kernel remembers information from cells executed earlier or out of order. Before trusting or exporting a notebook, restart the kernel and execute all cells from beginning to end. This checks that every result follows from the visible notebook content rather than from an earlier experiment.
</p>
</div>


Jupyter Notebook has two interaction modes. In <strong>edit mode</strong>, the cursor is inside a cell and you can edit its contents. In <strong>command mode</strong>, keyboard shortcuts operate on the selected cell as a whole. Press <code>Esc</code> to enter command mode and <code>Enter</code> to return to edit mode. The following table summarizes the most useful shortcuts:

<table style="
  width:65%;
  border-collapse:collapse;
  table-layout:fixed;
  margin:10px auto;
  color:black;
  text-align:left;
  font-size:95%;
">
  <colgroup>
    <col style="width:30%;">
    <col style="width:70%;">
  </colgroup>
  <thead>
    <tr style="background-color:#E3E3E3;">
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Shortcut</th>
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Action</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Shift + Enter</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Execute the selected cell and continue to the next cell.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Ctrl + Enter</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Execute the selected cell and keep it selected.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Esc</code> / <code>Enter</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Switch between command mode and edit mode.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>A</code> / <code>B</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Insert a new cell above or below the selected cell.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>M</code> / <code>Y</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Change the selected cell to Markdown or code.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>D</code>, <code>D</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Delete the selected cell by pressing <code>D</code> twice.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Z</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Undo the deletion of a cell.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Ctrl + S</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Save the notebook.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>Ctrl + Shift + H</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Display the available keyboard shortcuts.</td>
    </tr>
  </tbody>
</table>

Shortcuts such as <code>A</code>, <code>B</code>, <code>M</code>, and <code>Y</code> work only in <strong>command mode</strong>. If pressing one of these keys inserts a letter into the cell, you are still in edit mode. Press <code>Esc</code> and try again. All important actions are also available through the Jupyter menus and buttons, so using keyboard shortcuts is helpful but not required. Press <code>Ctrl + Shift + H</code> or open the <strong>Help</strong> menu to see the shortcuts available in your Jupyter installation.

Save your notebook regularly using <code>Ctrl + S</code> or the corresponding menu command. Closing the browser tab does not necessarily stop Jupyter Notebook or the running kernel. To end the session, save your notebook and return to the terminal from which Jupyter was started. Press <code>Ctrl + C</code> and follow the displayed instructions to stop the application. Further information is available in the official <a href="https://jupyter-notebook.readthedocs.io/en/latest/">Jupyter Notebook documentation</a>.

<a id="export"></a>
## Exporting Notebooks

A Jupyter notebook is an interactive document whose code cells can be edited and executed. It can also be exported as a **static document** that records its current text, code, figures, and saved results. An **HTML export** can be opened in a web browser and is well suited for online reading, while a **PDF export** is convenient for printing and sharing. The exported document can be viewed but its code cannot be executed interactively. You should therefore keep the original `.ipynb` file as the editable and executable version.

Before exporting a notebook, save it, restart the kernel, and execute all cells from beginning to end. This checks that the displayed results can be reproduced from the notebook itself and do not depend on cells executed earlier or out of order. To export through Jupyter Notebook, select <strong>File &rarr; Save and Export Notebook As</strong> and choose the desired format. Notebooks can also be processed from a terminal using [**nbconvert**](https://nbconvert.readthedocs.io/), a command-line tool provided by Jupyter. The following commands use the PCP overview notebook `PCP.ipynb` as an example. Run them from the PCP directory with the PCP environment active.

<table style="
  width:88%;
  border-collapse:collapse;
  table-layout:fixed;
  margin:10px auto;
  color:black;
  text-align:left;
  font-size:95%;
">
  <colgroup>
    <col style="width:50%;">
    <col style="width:50%;">
  </colgroup>
  <thead>
    <tr style="background-color:#E3E3E3;">
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Command</th>
      <th style="border:1px solid #CCCCCC; padding:4px 6px; text-align:left;">Purpose</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>jupyter nbconvert --to notebook --execute --inplace PCP.ipynb</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Execute all cells in a fresh kernel and save the new results in the notebook.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>jupyter nbconvert --to html PCP.ipynb</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Export the notebook and its saved results as an HTML file.</td>
    </tr>
    <tr style="background-color:white;">
      <td style="border:1px solid #CCCCCC; padding:4px 6px;"><code>jupyter nbconvert --to webpdf PCP.ipynb</code></td>
      <td style="border:1px solid #CCCCCC; padding:4px 6px;">Export the notebook and its saved results as a PDF file.</td>
    </tr>
  </tbody>
</table>

The following exercises complete the setup and introduce the basic workflow used throughout PCP. First, you will verify that Python and the required packages are available in the PCP environment. You will then perform a small notebook experiment, modify it, and confirm that it runs correctly from beginning to end. The goal is not yet to write Python code independently, but to become comfortable with the tools needed for the rest of the course.

<a id="exercise_setup"></a>

<div style="
  background-color: #B2EBF290;
  border: 1px solid #4DD0E190;
  padding: 10px 20px;
  margin: 10px 0;
">
<strong style="color:#006064; display:block; break-after:avoid; page-break-after:avoid;">Exercise 1: Verifying Your PCP Setup</strong>

<p style="color:#006064;">
This exercise guides you through the complete startup procedure and helps you verify that PCP is working on your computer. Because paths, version numbers, and program output depend on the individual installation, there is no single sample solution. If something does not work immediately, read the error message carefully and return to the corresponding section of this unit.
</p>

<ul style="color:#006064; margin-bottom:0;">
  <li>Open the Anaconda Prompt on Windows or a terminal application on macOS or Linux. Change to the PCP project directory and confirm that it contains <code>PCP.ipynb</code> and <code>environment.yml</code>.</li>
  <li>Use <code>conda info --envs</code> to verify that an environment named <code>PCP</code> exists. If it does not, create it using the supplied environment file.</li>
  <li>Activate the environment with <code>conda activate PCP</code>. What change do you observe in the terminal prompt?</li>
  <li>Use <code>conda list</code> to inspect the installed packages. Find the entries for Python, NumPy, Matplotlib, and Jupyter, and compare their versions with the requirements in <code>environment.yml</code>.</li>
  <li>Start Jupyter Notebook from the PCP directory and open <code>PCP.ipynb</code>. Follow the link to Unit 1 and confirm that the notebook opens correctly.</li>
  <li>Execute the import cell near the beginning of this unit. If it finishes without an error message, Python can find the main packages required by PCP.</li>
  <li>In a new code cell, display <code>sys.executable</code>, <code>sys.version</code>, <code>np.__version__</code>, and <code>matplotlib.__version__</code>. What does each value tell you about the active Python environment?</li>
  <li>Restart the kernel and execute the notebook from the beginning. Confirm that all cells up to this exercise run in their visible order without errors.</li>
</ul>
</div>

<a id="exercise_notebook"></a>

<div style="
  background-color: #B2EBF290;
  border: 1px solid #4DD0E190;
  padding: 10px 20px;
  margin: 10px 0;
">
<strong style="color:#006064; display:block; break-after:avoid; page-break-after:avoid;">Exercise 2: Your First Notebook Experiment</strong>

<p style="color:#006064;">
Now that PCP is running, explore the basic workflow of a Jupyter notebook. This exercise does not require previous knowledge of Python. Begin with the supplied code, observe its result, and then change it. Because you will choose your own parameters and write your own observations, there is no single sample solution. Use the code cell labeled <code># Your Solution</code> directly below this exercise. NumPy and Matplotlib were imported near the beginning of the notebook and are available under the names <code>np</code> and <code>plt</code>. Copy the following code into the solution cell and execute it using <code>Shift + Enter</code>:    
</p>

<pre style="
  background-color:#f5f5f5;
  border:1px solid #d6d6d6;
  border-radius:3px;
  padding:10px;
  overflow-x:auto;
"><code style="
  display:block;
  background-color:transparent !important;
">sr = 100
duration = 1
frequency = 3
amplitude = 1

t = np.arange(int(duration * sr)) / sr
x = amplitude * np.sin(2 * np.pi * frequency * t)

plt.figure(figsize=(6, 2.5))
plt.plot(t, x)
plt.xlabel('Time (seconds)')
plt.ylabel('Amplitude')
plt.title('Sinusoid')
plt.grid(alpha=0.3)
plt.show()</code></pre>

<ul style="color:#006064; margin-bottom:0;">
  <li>Describe the resulting curve in your own words.</li>
  <li>Change <code>frequency</code> and execute the cell again. How does the number of oscillations change?</li>
  <li>Change <code>amplitude</code>. Which property of the plot changes, and which property remains unchanged?</li>
  <li>Try at least two further parameter combinations of your own. Before executing the cell, predict how the result will look, and then compare your prediction with the resulting plot.</li>
  <li>Insert a Markdown cell above your solution. Give your experiment a heading and briefly describe the effects of changing <code>frequency</code> and <code>amplitude</code>.</li>
  <li>Deliberately introduce a small error, for example by replacing <code>frequency</code> with a name that has not been defined. Execute the cell, read the error message, and then repair the code.</li>
  <li>Restart the kernel and execute the complete Unit 1 notebook from beginning to end. Confirm that the experiment still works when all cells are executed in their visible order.</li>
  <li>Save the notebook and export it as an HTML document. Open the exported file in a browser and identify one important difference between the HTML document and the original notebook.</li>
</ul>
</div>

In [3]:
# Your Solution

<div>
<a href="https://opensource.org/licenses/MIT"><img src="data_layout/PCP_License.png" width=100% style="float: right;" alt="PCP License"></a>
</div>